app/services/action_builder.rb

```diff
@@ -20,6 +20,15 @@ def build_action(extra_attrs = {})
   def previous_action
```

Comment:
A bit late to the party as this is already merged... and since I don't know much about Champaigns internals it might be superfluous. Anyway:

* This method does not _memoize_ the result, so the ```@previous_action``` instance variable should not be required a local one is enough.
* If there is no @previous_action then the loop is always executed completely even when a ```Action``` is found for the first connected page (though the DB is not hit). As soon as one action is found the loop should/could stop.
* ```Thing.where(conditions).first``` could be replaced by ```Thing.find_by(conditions)``` (https://github.com/bbatsov/rails-style-guide#find_by)

Have a great day.